experimental/hip/memory_pools.c

```diff
@@ -0,0 +1,326 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "experimental/hip/memory_pools.h"
+
+#include "experimental/hip/dynamic_symbols.h"
+#include "experimental/hip/hip_buffer.h"
+#include "experimental/hip/status_util.h"
+
+// NOTE: these are currently global for all devices; we could make
+// device-specific ones by malloc() and leaking (with LSAN note) unique string
+// values instead.
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_ALLOCATION_TRACKING
+static const char* IREE_HAL_HIP_DEVICE_LOCAL_POOL_RESERVED_ID =
+    "HIP pool: device-local reserved";
+static const char* IREE_HAL_HIP_OTHER_POOL_RESERVED_ID =
+    "HIP pool: other reserved";
+#endif  // IREE_TRACING_FEATURE_ALLOCATION_TRACKING
+
+static iree_status_t iree_hal_hip_create_memory_pool(
+    const iree_hal_hip_dynamic_symbols_t* hip_symbols, hipDevice_t hip_device,
+    iree_hal_hip_memory_pool_params_t params,
+    hipMemPool_t* IREE_RESTRICT out_pool) {
+  *out_pool = NULL;
+
+  hipMemPoolProps pool_props = {
+      .allocType = hipMemAllocationTypePinned,
+      // TODO: allow sharing of certain pool memory types by fd/HANDLE.
+      .handleTypes = hipMemHandleTypeNone,
+      .location =
+          {
+              .type = hipMemLocationTypeDevice,
+              .id = hip_device,
+          },
+      .reserved = {0},
+      .win32SecurityAttributes = NULL,
+  };
+
+  hipMemPool_t pool = NULL;
+
+  // WARNING: hipMemPoolCreate() API is marked as beta in HIP library meaning
```

Comment:
We can have one comment in the `memory_pools.h` regarding the beta status; don't need to duplicate the message for every hip pool api call.